### DISCUSSION TITLE
Eglot sorts initial completions by length, not relevance.  When non-empty pattern given

### DISCUSSION THREAD

**Discussion by MintSoup** [joaotavora/eglot#1306] [Q&A] [answered]
I think it's best to illustrate with a picture:
![2023-10-16_12-24](https://github.com/joaotavora/eglot/assets/15715189/88e132de-5293-403d-8f22-e2111ca8bfad)
Notice how `aLongName1234` is displayed after `auto`, `abort`, etc, even though it's a local variable and should logically be displayed earlier in the list of completions (and does get displayed earlier with lsp-mode). 
![2023-10-16_12-25](https://github.com/joaotavora/eglot/assets/15715189/9943c601-bdee-4896-ab10-8d5fb32f07d4)

Interestingly enough, subsequent completions (if, for example, I type `aLongName1234.`) do seem to get sorted correctly.  Looks like eglot does client-side sorting sometimes. Any ideas?

**Comment by MintSoup** [comment]
Initially I thought this was a clangd-only issue but now I think I can reproduce it with `rust-analyzer` too.

**Comment by zsxh** [comment]
Check `completion-category-defaults`, eglot use `(styles flex basic)` completion style by default, you can customizable it via `completion-category-overrides`.

**Reply by joaotavora** [reply]
~@MintSoup  please ignore this advice: "flex" is only used as a implementation detail to do some highlighting of completions.~

EDIT: What I wrote is wrong.

**Reply by joaotavora** [reply]
Never mind, this advice by @zsxh is indeed correct.  My sincere apologies, I thought `flex` was used only for highlighting (which it is).  But it unfortunately also kicks in for overriding the sorting given by the completion table, which is based on the server's score.

**Comment by joaotavora** [comment]
I can't reproduce this @MintSoup.  You haven't provided any details of your configuration, and I recommend that you start with a bare-bones setup, i.e. a `Emacs -Q` recipe, with nothing but Eglot, clangd and company (I do hope you're using Company and not Corfu, as the latter is known to be problematic in more than one way).

Anyway, here's my `Emacs -Q` recipe:

```
HOME=`mktemp -d` ~/Source/Emacs/emacs/src/emacs --eval '(package-install (quote company))' -f global-company-mode /tmp/thingy.cpp -f eglot
```

Where `thingy.cpp` is just this:

```
#include <iostream>

int main() {
  std
```

Inserting `::` starts a Company completion session. If you like screenshots, here is one:

![image](https://github.com/joaotavora/eglot/assets/387011/59d61126-436c-4b62-b028-b57aa71db8e3)

As you can see, they are _not_ sorted according to length, but rather to the `:score` that the `clangd` assigns to them them

These are the labels and the scores that clangd assigns to each completion.

```
((" execution" . 1.320576786994934)
 ("•any" . 0.631610631942749)
 ("•bad_any_cast" . 0.631610631942749)
 ("•pointer_safety" . 0.631610631942749)
 ("•gslice" . 0.5928206443786621) ("•gslice_array<class Tp>" . 0.5928206443786621) ("•mask_array<class Tp>" . 0.5928206443786621) ("•priority_queue<typename Tp, typename Sequence, typename Compare>" . 0.5928206443786621) ("•queue<typename Tp, typename Sequence>" . 0.5928206443786621) ("•set<typename Key, typename Compare, typename Alloc>" . 0.5928206443786621) ("•uniform_int_distribution<typename IntType>" . 0.5928206443786621) ("•any_cast(…)" . 0.47691652178764343) ...)
```

**Reply by KaranAhlawat** [reply]
> I do hope you're using Company and not Corfu, as the latter is known to be problematic in more than one way

Interesting, is there some where I can read up on these problems? Since I personally haven't run into any so far

**Reply by joaotavora** [reply]
This is out of topic, but the problems are:

1. Occasional out-of-sync errors
2. Missing completions

See https://github.com/joaotavora/eglot/discussions/1127  for lots of scattered details.

There are a number of workarounds that involve monkey-patching Eglot in the Corfu website.  Presumably people get away with those, anyway but I don't provide support for monkey-patched Eglot.

**Reply by MintSoup** [reply]
> I do hope you're using Company

I am, but the issue persists with corfu and even the builtin `complete-symbol` too.

Here's a screenshot from an `emacs -q` session, demonstrating the same problem, without any third-party packages:
![2023-10-16_18-54](https://github.com/joaotavora/eglot/assets/15715189/c660b939-cb7d-468d-a539-b96a51a0dbda)

I just typed `a` and pressed `C-M-i`. Pure vanilla emacs 29, no external elisp loaded at all.

The same issue with rust-analyzer:
![2023-10-16_18-57](https://github.com/joaotavora/eglot/assets/15715189/55e3f94e-3087-4d90-a8ef-780008ef5f60)

Whereas with lsp-mode, `longName` shows up first in the list of completions:

![2023-10-16_18-59](https://github.com/joaotavora/eglot/assets/15715189/116d4054-8ec6-4081-a523-633dc51c3bbb)

Coincidentally, in cases when this ordering issue occurs, the completion popup seems to appear a few hundreds of milliseconds later than usual.

**Comment by joaotavora** [comment]
OK, I reproduced the issue with your example file.

Incidentally, I have to apologize to @zsxh, as their advice was right all along.  A workaround to fix this is to remove `eglot`'s entry from `completion-category-defaults` ,for example with:

```
(setf (alist-get 'eglot completion-category-defaults) nil)
```

This is a bug in Eglot.  `flex` is useful to highlight the parts of the completion that match the pattern (you'll notice Eglot does this and lsp-mode apparently doesn').  However, it has the side effect of overriding the sorting as well, which is unfortunate.  I'll see what I can do.

**Reply by MintSoup** [reply]
> OK, I reproduced the issue with your example file.

Great, this and #1141 were the only things keeping me from switching to eglot from lsp-mode. Once this gets fixed, I'll give it another try.

> Incidentally, I have to apologize to @zsxh, as their advice was right all along.

Ah, it's a shame I didn't even bother trying it before continuing.

**Comment by joaotavora** [comment, marked as answer]
I fixed this in the Emacs master branch.  For other Emacs versions, will be in the upcoming Eglot 1.16 release, but you can find it in [GNU devel ELPA](https://elpa.gnu.org/devel/eglot.html) before that.
